universum/modules/structure_handler.py

```diff
@@ -197,15 +198,22 @@ def process_one_step(self, merged_item: Step, step_executor: Callable, skip_exec
             self.report_skipped_block(numbering + "'" + merged_item.name + "'")
             return True
 
+        process = None
+        executed_successfully: bool = True
         # Here pass_errors=False, because any exception while executing build step
         # can be step-related and may not affect other steps
         with self.block(block_name=step_label, pass_errors=False):
-            error: Optional[str] = self.execute_one_step(merged_item, step_executor)
-            if error is not None:
+            process= self.execute_one_step(merged_item, step_executor)
+            error = process.get_error()
+            executed_successfully = (error is None)
+            if not executed_successfully:
+                error = error if error else ""
                 self.fail_current_block(error)
-                return False
+        if not merged_item.background:
+            with self.block(block_name=f"Collecting artifacts for the '{merged_item.name}' step", pass_errors=False):
```

Comment:
This approach creates blocks for cases when step doesn't declare any artifacts.
This is in contrast with current approach: block are created only for artifacts that are declared and if two or more steps declare the same artifact, the block is only created once.
I suggest to change the code to only start the collection if the step declares some artifacts.